python_modules/dagster/dagster/_core/definitions/cacheable_assets.py

```diff
@@ -0,0 +1,97 @@
+
+
+@whitelist_for_serdes
+class AssetsDefinitionMetadata(
+    NamedTuple(
+        "_AssetsDefinitionMetadata",
+        [
+            ("keys_by_input_name", Optional[Mapping[str, AssetKey]]),
+            ("keys_by_output_name", Optional[Mapping[str, AssetKey]]),
+            ("internal_asset_deps", Optional[Mapping[str, AbstractSet[AssetKey]]]),
+            ("group_name", Optional[str]),
+            ("metadata_by_output_name", Optional[Mapping[str, MetadataUserInput]]),
+            ("key_prefix", Optional[CoercibleToAssetKeyPrefix]),
+            ("can_subset", bool),
+            ("extra_metadata", Optional[Mapping[Any, Any]]),
+        ],
+    )
+):
+    """Data representing cacheable metadata about assets, which can be used to generate
+    AssetsDefinition objects in other processes
+    """
+
+    def __new__(
+        cls,
+        keys_by_input_name: Optional[Mapping[str, AssetKey]] = None,
+        keys_by_output_name: Optional[Mapping[str, AssetKey]] = None,
+        internal_asset_deps: Optional[Mapping[str, AbstractSet[AssetKey]]] = None,
+        group_name: Optional[str] = None,
+        metadata_by_output_name: Optional[Mapping[str, MetadataUserInput]] = None,
+        key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
+        can_subset: bool = False,
+        extra_metadata: Optional[Mapping[Any, Any]] = None,
+    ):
+
+        extra_metadata = check.opt_nullable_mapping_param(extra_metadata, "extra_metadata")
+        try:
+            # check that the value is JSON serializable
+            seven.dumps(extra_metadata)
+        except TypeError:
+            check.failed("Value for `extra_metadata` is not JSON serializable.")
+
+        return super().__new__(
+            cls,
+            keys_by_input_name=check.opt_nullable_mapping_param(
+                keys_by_input_name, "keys_by_input_name", key_type=str, value_type=AssetKey
+            ),
+            keys_by_output_name=check.opt_nullable_mapping_param(
+                keys_by_output_name, "keys_by_output_name", key_type=str, value_type=AssetKey
+            ),
+            internal_asset_deps=check.opt_nullable_mapping_param(
+                internal_asset_deps, "internal_asset_deps", key_type=str, value_type=set
+            ),
+            group_name=check.opt_str_param(group_name, "group_name"),
+            metadata_by_output_name=check.opt_nullable_mapping_param(
+                metadata_by_output_name, "metadata_by_output_name", key_type=str, value_type=dict
+            ),
+            key_prefix=check.opt_inst_param(key_prefix, "key_prefix", (str, list)),
+            can_subset=check.opt_bool_param(can_subset, "can_subset", default=False),
+            extra_metadata=extra_metadata,
+        )
+
+    def __hash__(self):
+        """Some of the values of this NamedTuple may be recursive dictionaries which are not
+        hashable. Instead, we serialize this as a namedtuple and hash that string.
+        """
```

Comment:
recursive freeze the dicts? serializing is not cheap when this data structure gets big